### ISSUE TITLE
method dispatch ambiguity failure

### ISSUE THREAD

**Issue by vtjnash** [JuliaLang/julia#8652]
The following ambiguous method situation fails to produce a method ambiguity warning:

```
~/Documents/no-backup/julia$ ./julia 
julia> f{T<:Integer}(::Type{T}, ::Ptr{Integer}) = 2
f (generic function with 1 method)

julia> f{T<:Integer}(::Type{T}, ::Ptr{T}) = 1
f (generic function with 2 methods)

julia> f(Integer, convert(Ptr{Integer}, C_NULL))
1

~/Documents/no-backup/julia$ ./julia 
julia> f{T<:Integer}(::Type{T}, ::Ptr{T}) = 1
f (generic function with 1 method)

julia> f{T<:Integer}(::Type{T}, ::Ptr{Integer}) = 2
f (generic function with 2 methods)

julia> f(Integer, convert(Ptr{Integer}, C_NULL))
2
```

this affects the behavior of the following lines:
https://github.com/JuliaLang/Gtk.jl/blob/5d1e5ee360aafae97c2189b8970345b70707024a/src/GLib/gtype.jl#L270
https://github.com/JuliaLang/Gtk.jl/blob/5d1e5ee360aafae97c2189b8970345b70707024a/src/GLib/gtype.jl#L284
currently, it works as intended, but if I switch the order of those two statements, it instead turns into a stack overflow

**Comment by timholy** [comment]
Fixed in #11194 (if that gets merged).

**Comment by timholy** [comment]
I thought I had a fix in deb0fb287d8a6bceea512bb56863afabbb782ea0, but that broke this rather fundamental property:
```
sptest4{T}(x::T, y::T) = 42
sptest4{T}(x::T, y) = 44

julia> @test sptest4(1,2) == 42

julia> @test sptest4(1, "cat") == 44
```

My "fix" made assumptions about how things are supposed to work based on the relationships among certain tests, but there's enough inconsistency that I think we first need to decide on the proper behavior:
```jl
julia> args_morespecific(a, b) = ccall(:jl_args_morespecific, Cint, (Any,Any), a, b)
args_morespecific (generic function with 1 method)

julia> T1 = TypeVar(:T, Integer, true); T2 = TypeVar(:T, Integer, true)
T<:Integer

julia> a = Tuple{T1, T1}
Tuple{T<:Integer,T<:Integer}

julia> b1 = Tuple{T1, Integer}
Tuple{T<:Integer,Integer}

julia> b2 = Tuple{T2, Integer}
Tuple{T<:Integer,Integer}

julia> args_morespecific(a, b1)
0

julia> args_morespecific(b1, a)
0

julia> args_morespecific(a, b2)
1

julia> args_morespecific(b2, a)
0

julia> a = Tuple{T1, Ptr{T1}}
Tuple{T<:Integer,Ptr{T<:Integer}}

julia> b1 = Tuple{T1, Ptr{Integer}}
Tuple{T<:Integer,Ptr{Integer}}

julia> b2 = Tuple{T2, Ptr{Integer}}
Tuple{T<:Integer,Ptr{Integer}}

julia> args_morespecific(a, b1)
0

julia> args_morespecific(b1, a)
1

julia> args_morespecific(a, b2)
1

julia> args_morespecific(b2, a)
1
```
In any real method-sorting test, it's the comparison of `a` and `b2` that matters, since two separate methods won't share a TypeVar. I'll proceed on the assumption that the inconsistencies in the `a`, `b1` tests don't matter.
